Broken generated pnpm lock file (ERR_PNPM_ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY)

### Current Behavior

I'm using `@nx/webpack:webpack` executor with `"generatePackageJson": true`.

When I'm building production build of app and deploying it I'm getting following error:

```
Lockfile is up to date, resolution step is skipped
 WARN  Broken lockfile: no entry for '/strip-ansi/6.0.1' in pnpm-lock.yaml
 ERR_PNPM_ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY  The lockfile is broken! Resolution step will be performed to fix it.
```



### Expected Behavior

lock file shoud be correct and there should be no error.

### GitHub Repo

https://github.com/diginikkari/nx-next

### Steps to Reproduce

1. add app with webpack plugin `pnpm nx g @nx/react:app my-app --bundler=webpack`
2. enable 'generatePackageJson' in ´project.json´
3. install `glob` as dev dependency `pnpm install -d glob@10.2.6` (this will add `string-width@5.1.2` dependency)
4. install `msw` as dependency `pnpm install msw` (this will add `string-width@4.2.3` dependency)
5. add `import { rest } from 'msw';` to app to include msw as dependency for app
6. build app `pnpm nx run my-app:build:production`
7. copy generated `dist` folder to outside of workspace and run 'pnpm install' inside it.

### Nx Report

```shell
Node   : 20.2.0
   OS     : darwin-arm64
   pnpm   : 8.6.1
   
   nx                 : 16.4.0-beta.2
   @nx/js             : 16.4.0-beta.2
   @nx/jest           : 16.4.0-beta.2
   @nx/linter         : 16.4.0-beta.2
   @nx/workspace      : 16.4.0-beta.2
   @nx/cypress        : 16.4.0-beta.2
   @nx/devkit         : 16.4.0-beta.2
   @nx/esbuild        : 16.4.0-beta.2
   @nx/eslint-plugin  : 16.4.0-beta.2
   @nx/next           : 16.4.0-beta.2
   @nx/node           : 16.4.0-beta.2
   @nx/react          : 16.4.0-beta.2
   @nx/storybook      : 16.4.0-beta.2
   @nrwl/tao          : 16.4.0-beta.2
   @nx/vite           : 16.4.0-beta.2
   @nx/web            : 16.4.0-beta.2
   @nx/webpack        : 16.4.0-beta.2
   nx-cloud           : 16.0.5
   typescript         : 5.1.3
   ---------------------------------------
   Community plugins:
   @mands/nx-playwright : 0.6.2
   ---------------------------------------
```


### Failure Logs

_No response_

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

I tried it with `16.3.2` and with latest beta `16.4.0-beta.2`.

Lockfile generation seems to select `/string-width@5.1.2` entry into the generated lock file even though there is only `string-width: 4.2.3` dependency in app's dependencies.

--- a/packages/nx/src/plugins/js/lock-file/pnpm-parser.ts
+++ b/packages/nx/src/plugins/js/lock-file/pnpm-parser.ts
@@ -18,7 +18,7 @@ import {
   ProjectGraph,
   ProjectGraphExternalNode,
 } from '../../../config/project-graph';
-import { fileHasher, hashArray } from '../../../hasher/file-hasher';
+import { hashArray } from '../../../hasher/file-hasher';
 
 export function parsePnpmLockfile(
   lockFileContent: string,
@@ -43,39 +43,40 @@ function addNodes(
   const nodes: Map<string, Map<string, ProjectGraphExternalNode>> = new Map();
 
   Object.entries(data.packages).forEach(([key, snapshot]) => {
-    const packageName = findPackageName(key, snapshot, data);
-    const rawVersion = findVersion(key, packageName);
-    const version = parseBaseVersion(rawVersion, isV6);
+    findPackageNames(key, snapshot, data).forEach((packageName) => {
+      const rawVersion = findVersion(key, packageName);
+      const version = parseBaseVersion(rawVersion, isV6);
 
-    // we don't need to keep duplicates, we can just track the keys
-    const existingNode = nodes.get(packageName)?.get(version);
-    if (existingNode) {
-      keyMap.set(key, existingNode);
-      return;
-    }
+      // we don't need to keep duplicates, we can just track the keys
+      const existingNode = nodes.get(packageName)?.get(version);
+      if (existingNode) {
+        keyMap.set(key, existingNode);
+        return;
+      }
 
-    const node: ProjectGraphExternalNode = {
-      type: 'npm',
-      name: version ? `npm:${packageName}@${version}` : `npm:${packageName}`,
-      data: {
-        version,
-        packageName,
-        hash:
-          snapshot.resolution?.['integrity'] ||
-          hashArray(
-            snapshot.resolution?.['tarball']
-              ? [snapshot.resolution['tarball']]
-              : [packageName, version]
-          ),
-      },
-    };
+      const node: ProjectGraphExternalNode = {
+        type: 'npm',
+        name: version ? `npm:${packageName}@${version}` : `npm:${packageName}`,
+        data: {
+          version,
+          packageName,
+          hash:
+            snapshot.resolution?.['integrity'] ||
+            hashArray(
+              snapshot.resolution?.['tarball']
+                ? [snapshot.resolution['tarball']]
+                : [packageName, version]
+            ),
+        },
+      };
 
-    keyMap.set(key, node);
-    if (!nodes.has(packageName)) {
-      nodes.set(packageName, new Map([[version, node]]));
-    } else {
-      nodes.get(packageName).set(version, node);
-    }
+      keyMap.set(key, node);
+      if (!nodes.has(packageName)) {
+        nodes.set(packageName, new Map([[version, node]]));
+      } else {
+        nodes.get(packageName).set(version, node);
+      }
+    });
   });
 
   const hoistedDeps = loadPnpmHoistedDepsDefinition();
@@ -280,11 +281,14 @@ function findVersion(key: string, packageName: string): string {
   return key;
 }
 
-function findPackageName(
+function findPackageNames(
   key: string,
   snapshot: PackageSnapshot,
   data: Lockfile
-): string {
+): string[] {
+  const packageNames = new Set<string>();
+  const originalPackageName = extractNameFromKey(key);
+
   const matchPropValue = (record: Record<string, string>): string => {
     if (!record) {
       return undefined;
@@ -292,6 +296,13 @@ function findPackageName(
     const index = Object.values(record).findIndex((version) => version === key);
     if (index > -1) {
       return Object.keys(record)[index];
+    }
+    // check if non aliased name is found
+    if (
+      record[originalPackageName] &&
+      key.startsWith(`/${originalPackageName}/${record[originalPackageName]}`)
+    ) {
+      return originalPackageName;
     }
   };
 
@@ -307,7 +318,7 @@ function findPackageName(
 
   // snapshot already has a name
   if (snapshot.name) {
-    return snapshot.name;
+    packageNames.add(snapshot.name);
   }
   // it'a a root dependency
   const rootDependencyName =
@@ -315,17 +326,20 @@ function findPackageName(
     // only root importers have devDependencies
     matchPropValue(data.importers['.'].devDependencies);
   if (rootDependencyName) {
-    return rootDependencyName;
+    packageNames.add(rootDependencyName);
   }
   // find a snapshot that has a dependency that points to this snapshot
   const snapshots = Object.values(data.packages);
   for (let i = 0; i < snapshots.length; i++) {
     const dependencyName = matchedDependencyName(snapshots[i]);
     if (dependencyName) {
-      return dependencyName;
+      packageNames.add(dependencyName);
     }
   }
-  return extractNameFromKey(key);
+  if (packageNames.size === 0) {
+    packageNames.add(originalPackageName);
+  }
+  return Array.from(packageNames);
 }
 
 function getVersion(key: string, packageName: string): string {

